### ISSUE TITLE
Annoy less resilient to id holes than before

### ISSUE THREAD

**Issue by yonromai** [spotify/annoy#223]
Hey @erikbern, 

TL;DR: `1.9.4` is a lot less robust to holes in the id space than `1.3.2`.

--

After upgrading annoy to `1.9.4` for DW, we got a pretty significant drop in metrics. 

After investigations, it seems that compared to older versions of annoy, the latest version is not very robust to holes in the IDs (which we have for DW because of some multistage filtering logic). 

In an experiment where I compared building a tree without holes (id: 0 to n-1) with 30% of holes in the ids, I found that annoy `1.3.2` got a 3% drop in perf (cos sim of nn) where annoy `1.9.4` perf dropped by 36%.

Here are the numbers for mean cos sim of a bunch of nn queries:
```
	old_annoy	new_annoy	old_holes_annoy	new_holes_annoy
count	5000.000000	5000.000000	3499.000000	3520.000000
mean	0.135056	0.132766	0.140376	0.180412
std	0.013602	0.013103	0.017448	0.017294
min	0.099506	0.096045	0.091715	0.130184
25%	0.125277	0.123374	0.128233	0.168452
50%	0.134288	0.131794	0.138153	0.179342
75%	0.143350	0.141121	0.150459	0.191101
max	0.197450	0.189082	0.218467	0.257536
```

(If you want the code to reproduce this I can open source it)

Were you aware of this issue?

**Comment by a1k0n** [comment]
That's weird. Are the "hole" nodes all zero vectors in both versions if you query them directly?

**Comment by yonromai** [comment]
Yes, that's a good point. So perhaps the drop I'm seeing is due to a bunch of all zero vecs polluting the results? or perhaps just the increase in tree size?

**Comment by erikbern** [comment]
i never really designed annoy to care about holes – it's been working somewhat unintentionally.

should be possible to filter them out on index building pretty easily. look at https://github.com/spotify/annoy/blob/master/src/annoylib.h#L526 – all you need to do there is to ignore any vector that's all zeros.

**Comment by erikbern** [comment]
let me know if you aren't able to fix it – i could probably put together a quick fix this weekend

**Comment by yonromai** [comment]
Thanks for coming back to me. Next week's crazy so I won't have much time to look into it but don't worry too much about it; we've been using a stale version of annoy for 2 years, we can wait a bit more xD

**Comment by a1k0n** [comment]
What I meant by my comment is that the holes might actually be uninitialized memory, and not actually zero vectors. You could be getting junk data in the new version but zeros in the old, or something. I'm suggesting you actually check the index files to see.

**Comment by erikbern** [comment]
@a1k0n should be easy to memzero or bzero or whatever the function is called...i thought i already did that but could be wrong

**Comment by a1k0n** [comment]
Actually yeah, nevermind. `add_item` calls `_allocate_size` which does a memset, but even if it didn't, there's no API to just add a "hole"... you have to add all the vectors in order, so unused items will have whatever vectors you give them... @yonromai what is your `_holes_*` code doing?

**Comment by erikbern** [comment]
Ok so I don't see any *bugs* with the code at least. This works:

```python
class HolesTest(TestCase):
    def test_holes(self):
	f = 10
        index = AnnoyIndex(f)
        valid_indices = set()
	for i in range(10000):
            i2 = int(i*2**-0.5) # leave holes every few items                                                                                                                                                                              
	    valid_indices.add(i2)
            v = numpy.random.normal(size=(f,))
            index.add_item(i2, v)
	index.build(10)
        for i in valid_indices:
            js = index.get_nns_by_item(i, 100)
            for j in js:
                self.assertTrue(j in valid_indices)
        for i in range(1000):
	    v = numpy.random.normal(size=(f,))
	    js = index.get_nns_by_vector(v, 100)
            for j in js:
                self.assertTrue(j in valid_indices)
```

I'll take a look and see if I can improve the index building though

**Comment by erikbern** [comment]
actually nvm i can reproduce this

```python
class HolesTest(TestCase):
    def test_holes(self):
	f = 10
        index = AnnoyIndex(f)
        index.add_item(1000, numpy.random.normal(size=(f,)))
	index.build(10)
	js = index.get_nns_by_vector(numpy.random.normal(size=(f,)), 100)
	self.assertEquals(js, [1000])
```
